File: app/src/main/java/de/michael/filebinmobile/MainActivity.java
package de.michael.filebinmobile;

import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v4.app.Fragment;
import android.support.v4.app.FragmentManager;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.view.MenuItem;

import butterknife.BindView;
import butterknife.ButterKnife;
import de.michael.filebinmobile.controller.SettingsManager;
import de.michael.filebinmobile.fragments.HistoryFragment;
import de.michael.filebinmobile.fragments.PasteFragment;
import de.michael.filebinmobile.fragments.SettingsFragment;

public class MainActivity extends AppCompatActivity implements OnTabNavigationRequestedListener {

    @BindView(R.id.bnvMainNavigation)
    BottomNavigationView bnvMainNavigation;

    // TODO remove global vars and try to retrieve fragments by id
    private Fragment pasteFragment, historyFragment, settingsFragment;

    private BottomNavigationView.OnNavigationItemSelectedListener onNavigationItemSelectedListener
            = new BottomNavigationView.OnNavigationItemSelectedListener() {

        @Override
        public boolean onNavigationItemSelected(@NonNull MenuItem item) {
            switch (item.getItemId()) {
                case R.id.navigation_paste:
                    if (pasteFragment == null) {

                        pasteFragment = new PasteFragment();
                        //TODO create a super class for our fragments and set a listener there
                        ((PasteFragment) pasteFragment).setOnTabNavigationRequestedListener(MainActivity.this);

                    }

                    switchFragment(pasteFragment);
                    return true;
                case R.id.navigation_history:
                    if (historyFragment == null) {

                        historyFragment = new HistoryFragment();

                    }
                    switchFragment(historyFragment);
                    return true;
                case R.id.navigation_server_settings:
                    if (settingsFragment == null) {

                        settingsFragment = new SettingsFragment();

                    }
                    switchFragment(settingsFragment);
                    return true;
            }
            return false;
        }
    };

    private void switchFragment(Fragment fragment) {
        FragmentManager manager = getSupportFragmentManager();

        // do not add any transitions of bottom bar navigation views to back stack
        // as stated in android design guidelines. See link below. However we will implement the
        // behaviour for pressing back as returning to our "home" view, the first menu item (Paste)
        // https://material.io/guidelines/components/bottom-navigation.html#bottom-navigation-behavior
        manager.beginTransaction().replace(R.id.frlMainContent, fragment).commit();
    }

    @Override
    public void onBackPressed() {

        int selectedItemId = bnvMainNavigation.getSelectedItemId();

        if (selectedItemId != R.id.navigation_paste) {
            bnvMainNavigation.setSelectedItemId(R.id.navigation_paste);
        } else {
            super.onBackPressed();
        }
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        ButterKnife.bind(this);

        this.pasteFragment = new PasteFragment();
        ((PasteFragment) pasteFragment).setOnTabNavigationRequestedListener(MainActivity.this);

        // don't add the first transaction the the backstack so we don't have this weird empty view
        getSupportFragmentManager().beginTransaction()
                .add(R.id.frlMainContent, this.pasteFragment).commit();

        bnvMainNavigation.setOnNavigationItemSelectedListener(this.onNavigationItemSelectedListener);

        // Do some stuff on first launch
        boolean hasLaunchedBefore = SettingsManager.getInstance().hasLaunchedBefore(this);
        if (!hasLaunchedBefore) {

            onFirstLaunch();

        }

    }

    /**
     * We wan't to show a welcome screen and give a short introduction on adding server profiles
     */
    public void onFirstLaunch() {

        new AlertDialog.Builder(this)
                .setTitle(R.string.welcome)
                .setMessage(R.string.welcomeMsg)
                .setPositiveButton(R.string.navToServerSettings, (dialogInterface, i) ->
                        onNavigationRequest(R.id.navigation_server_settings))
                .setNegativeButton(R.string.no, (dialogInterface, i) -> dialogInterface.dismiss())
                .create().show();


    }

    @Override
    public void onNavigationRequest(int menuItemId) {
        bnvMainNavigation.setSelectedItemId(menuItemId);
    }
}

File: app/src/main/java/de/michael/filebinmobile/MainActivity.kt
package de.michael.filebinmobile

import android.os.Bundle
import android.view.MenuItem
import androidx.appcompat.app.AlertDialog
import androidx.appcompat.app.AppCompatActivity
import androidx.fragment.app.Fragment
import de.michael.filebinmobile.controller.SettingsManager
import de.michael.filebinmobile.fragments.HistoryFragment
import de.michael.filebinmobile.fragments.PasteFragment
import de.michael.filebinmobile.fragments.SettingsFragment
import kotlinx.android.synthetic.main.activity_main.*

class MainActivity : AppCompatActivity() {

    private val onNavigationRequest: (Int) -> Unit = {
        bnvMainNavigation.selectedItemId = it
    }

    private val onNavItemSelected = { item: MenuItem ->
        when (item.itemId) {
            R.id.navigation_paste -> {
                val pasteFragment = PasteFragment()
                pasteFragment.onTabNavigationRequested = onNavigationRequest
                switchFragment(pasteFragment)
                true

            }
            R.id.navigation_history -> {
                val historyFragment = HistoryFragment()
                historyFragment.onTabNavigationRequested = onNavigationRequest
                switchFragment(historyFragment)
                true
            }
            R.id.navigation_server_settings -> {
                val settingsFragment = SettingsFragment()
                settingsFragment.onTabNavigationRequested = onNavigationRequest
                switchFragment(settingsFragment)
                true
            }
            else -> false
        }
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        // add our "home"/main fragment
        val pasteFragment = PasteFragment()
        // add our navigation function so we can send navigation requests from inside the fragment
        pasteFragment.onTabNavigationRequested = onNavigationRequest
        switchFragment(pasteFragment)

        // navigation listener for our navigation bar at the bottom
        bnvMainNavigation.setOnNavigationItemSelectedListener(onNavItemSelected)

        // First launch?
        if (SettingsManager.hasLaunchedBefore(this)) {
            onFirstLaunch()
            SettingsManager.setHasLaunchedBefore(this, true)
        }
    }

    override fun onBackPressed() {
        // return to our "home" view or exit out of the app
        // behaviour according to android design guidelines
        if (bnvMainNavigation.selectedItemId != R.id.navigation_paste) {
            onNavigationRequest(R.id.navigation_paste)
        } else {
            super.onBackPressed()
        }
    }

    private fun onFirstLaunch() {
        AlertDialog.Builder(this)
                .setTitle(R.string.welcome)
                .setMessage(R.string.welcomeMsg)
                .setPositiveButton(R.string.navToServerSettings) { _, _ ->
                    onNavigationRequest(R.id.navigation_server_settings)
                }
                .setNegativeButton(R.string.no) { dialogInterface, _ ->
                    dialogInterface.dismiss()
                }
                .create()
                .show()
    }

    private fun switchFragment(fragment: Fragment) {

        // do not add any transitions of bottom bar navigation views to back stack
        // as stated in android design guidelines. See link below. However we will implement the
        // behaviour for pressing back as returning to our "home" view, the first menu item (Paste)
        // https://material.io/guidelines/components/bottom-navigation.html#bottom-navigation-behavior

        supportFragmentManager.beginTransaction()
            .replace(R.id.frlMainContent, fragment)
            .commit()
    }
}